taichi/backends/vulkan/spirv_ir_builder.cpp

```diff
@@ -0,0 +1,940 @@
+    Value _1000000007u = uint_immediate_number(t_uint32_, 1000000007u);
+    Value _362436069u = uint_immediate_number(t_uint32_, 362436069u);
+    Value _521288629u = uint_immediate_number(t_uint32_, 521288629u);
+    Value _88675123u = uint_immediate_number(t_uint32_, 88675123u);
+    Value _1 = int_immediate_number(t_int32_, 1);
+    Value _1024 = int_immediate_number(t_int32_, 1024);
+
+    // init_rand_ segment (inline to main)
+    // ad-hoc: hope no kernel will use more than 1024 gtmp variables...
+    ib_.begin(spv::OpAccessChain).add_seq(gtmp_type, rand_gtmp_, 
+        global_tmp_, const_i32_zero_, _1024).commit(&func_header_);
+    // Get gl_GlobalInvocationID.x, assert it has be visited 
+    // (in generate_serial_kernel/generate_range_for_kernel
+    SType pint_type = this->get_pointer_type(t_uint32_, spv::StorageClassInput);
+    Value tmp0 = new_value(pint_type, ValueKind::kVariablePtr);
+    ib_.begin(spv::OpAccessChain).add_seq(pint_type, tmp0, gl_global_invocation_id,
+                              uint_immediate_number(t_uint32_, 0)).commit(&func_header_);
+    Value tmp1 = load_var(tmp0, t_uint32_);
+    Value tmp2_ = load_var(rand_gtmp_, t_int32_);
+    Value tmp2 = new_value(t_uint32_, ValueKind::kNormal);
+    ib_.begin(spv::OpBitcast).add_seq(t_uint32_, tmp2, tmp2_).commit(&func_header_);
+    Value tmp3 = new_value(t_uint32_, ValueKind::kNormal);
+    ib_.begin(spv::OpIAdd).add_seq(t_uint32_, tmp3, _7654321u, tmp1).commit(&func_header_);
+    Value tmp4 = new_value(t_uint32_, ValueKind::kNormal);
+    ib_.begin(spv::OpIMul).add_seq(t_uint32_, tmp4, _9723451u, tmp2).commit(&func_header_);
+    Value tmp5 = new_value(t_uint32_, ValueKind::kNormal);
+    ib_.begin(spv::OpIAdd).add_seq(t_uint32_, tmp5, _1234567u, tmp4).commit(&func_header_);
+    Value tmp6 = new_value(t_uint32_, ValueKind::kNormal);
+    ib_.begin(spv::OpIMul).add_seq(t_uint32_, tmp6, tmp3, tmp5).commit(&func_header_);
+    Value tmp7 = new_value(t_uint32_, ValueKind::kNormal);
+    ib_.begin(spv::OpIMul).add_seq(t_uint32_, tmp7, _123456789u, tmp6).commit(&func_header_);
+    Value tmp8 = new_value(t_uint32_, ValueKind::kNormal);
+    ib_.begin(spv::OpIMul).add_seq(t_uint32_, tmp8, _1000000007u, tmp7).commit(&func_header_);
+    store_var(_rand_x_, tmp8);
+    store_var(_rand_y_, _362436069u);
+    store_var(_rand_z_, _521288629u);
+    store_var(_rand_w_, _88675123u);
+    // Yes, this is not an atomic operation, but just fine since no matter
+    // how RAND_STATE changes, `gl_GlobalInvocationID.x` can still help
+    // us to set different seeds for different threads.
+    // Discussion: https://github.com/taichi-dev/taichi/pull/912#discussion_r419021918
+    Value tmp9 = load_var(rand_gtmp_, t_int32_);
+    Value tmp10 = new_value(t_int32_, ValueKind::kNormal);
+    ib_.begin(spv::OpIAdd).add_seq(t_int32_, tmp10, tmp9, _1).commit(&func_header_);
+    store_var(rand_gtmp_, tmp10);
+
+    init_rand_ = true;
+  }
```

Comment:
same thing here (runtime should be in a glsl & precompiled)

Reply:
Ack. Let's start with this, then migrate to the GLSL-based helpers.